fix bugs of container cpu shares when cpu request set to zero

waynepeking348 opened the pull request:
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:
For container resources, if limits are specified, but requests are not, default requests are set to limits in API server, and kubelet repeat this logic for containers running on existing Kubernetes cluster, but it works with some mistake for cpu. 
Actually, if cpu request is set to zero (instead of not specified), kubelet will also consider it as equal to limits, and set cpu shares as to limit instead of 2 (minShares). 
This is unexpected, think that if we have 2 containers, one with requests set as:

    limits:
        cpu: "4"
      requests:
        cpu: "2"

and the other with requests set as:

    limits:
        cpu: "4"
      requests:
        cpu: "0"

the cpu shares (in container level cgroup) for first one would be set as 2048 while the latter as 4096
but the expected cpu shares would be 2048 and 2 instead

**Special notes for your reviewer**:

I have one PR before #100986, but it's too out-of-date, so I open a new one rebased on current master

**Does this PR introduce a user-facing change?**:

```release-note
fix relative cpu priority for pods where containers explicitly request zero cpu by giving the lowest priority instead of falling back to the cpu limit to avoid possible cpu starvation of other pods
```

waynepeking348 committed:
fix bugs of container cpu shares when cpu request set to zero

waynepeking348 commented:
@andrewsykim @odinuge would you please have a look at this fix? thx

yangjunmyfm192085 reviewed (commented):
/ok-to-test

yangjunmyfm192085 commented:
/triage accepted

waynepeking348 committed:
fix test cases for cpu shares

waynepeking348 commented:
@yangjunmyfm192085 hi~ those checks have all been passed, could you please help merge this pr?

odinuge reviewed (approved):
Thanks for creating this PR @waynepeking348!

This fixes a problem where a user could insert a "sky high" cpu limit and then explicitly set `0` as the request, and starve all other workloads!

I think we should back port this one as well, since it fixes a real problem that can cause system instability!

/priority important-longterm

Except for the minor nit; 
/lgtm

I think the changelog value should be rephrased to something like (tho. my English is not my native language, so happy for others to phrase it better);

```
Fix relative cpu priority for pods where containers explicitly request zero cpu by giving the lowest priority instead of falling back to the cpu limit to avoid possible cpu starvation of other pods
```

waynepeking348 committed:
keep original cpu limit logic to be in line with the comments

waynepeking348 commented:
/retest


<br class="Apple-interchange-newline">

waynepeking348 commented:
@yangjunmyfm192085 @odinuge @ehashman hi～ I changed code to in line with the comment

    if cpuRequest == nil && cpuLimit != nil {
		cpuShares = int64(cm.MilliCPUToShares(cpuLimit.MilliValue()))
	} else {
	
and change log is also changed according to @odinuge suggestion

I think I need both /lgtm and /appoved label to make it merged, thx~

odinuge reviewed (approved):
/lgtm

waynepeking348 commented:
> approved

@odinuge I think I still need a /approved label to make it merged

yangjunmyfm192085 reviewed (commented):
/cc @mrunalp

yangjunmyfm192085 reviewed (commented):
/assign @mrunalp

waynepeking348 commented:
> /assign @mrunalp

hi, I think @mrunalp has missed this info, and I have tried to contact with him in slack, but also without reply, can anyone help to merge this pr please? @derekwaynecarr @klueska @yujuhong

waynepeking348 commented:
/assign @derekwaynecarr @klueska @yujuhong

waynepeking348 commented:
/remove-lifecycle stale

waynepeking348 commented:
/milestone v1.25
